crates/red_knot_test/src/matcher.rs

```diff
@@ -136,7 +136,16 @@ trait UnmatchedWithColumn {
     fn unmatched_with_column(&self, column: OneIndexed) -> String;
 }
 
-impl Unmatched for Assertion<'_> {
+// This is necessary since we only parse assertions lazily,
+// and sometimes we know before parsing any assertions that an assertion will be unmatched,
+// e.g. if we've exhausted all diagnostics but there are still assertions left.
```

Comment:
Seems a little odd that you only get assertion-parsing errors if there is at least one diagnostic that could match the assertion, but I think in practice it's fine?

Reply:
Would it be hard to force-parse unmatched assertions?

Reply:
I don't think it fits neatly into the current architecture. I'm going to defer it for now.

Reply:
I'll add a TODO